diff --git a/Makefile.in b/Makefile.in
index 2470891..fb10d11 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -7,8 +7,8 @@ CC_STD = -std=c99
 FLEX = flex
 BISON = bison
 CPPFLAGS += $(CC_STD) -D_POSIX_C_SOURCE
-CFLAGS = -pedantic -g -Wall -O2
-LDFLAGS = -g -lgmp -lgmpxx -lstdc++
+CFLAGS = -pedantic -g -Wall -O2 -m32
+LDFLAGS = -g -lgmp -lgmpxx -lstdc++ -m32
 
 OBJECTS = smtlib2bisonparser.o \
 	  smtlib2flexlexer.o \
@@ -51,7 +51,7 @@ install: libsmtlib2parser.a $(DIST_HEADERS)
 
 
 smtlib2yices: libsmtlib2parser.a $(YICES_OBJECTS)
-	$(CC) $(YICES_OBJECTS) libsmtlib2parser.a $(YICES_DIR)/lib/libyices.a $(LDFLAGS) -o $@
+	$(CC) $(YICES_OBJECTS) libsmtlib2parser.a $(YICES_DIR)/lib/libyices.a $(LDFLAGS) -o $@ -lm
 
 smtlib2yices.o: smtlib2yices.c
 	$(CC) $(CFLAGS) -I$(YICES_DIR)/include -c -o $@ $<
diff --git a/smtlib2abstractparser.h b/smtlib2abstractparser.h
index 99f199b..4af1f0b 100644
--- a/smtlib2abstractparser.h
+++ b/smtlib2abstractparser.h
@@ -34,9 +34,21 @@
 
 typedef struct smtlib2_abstract_parser smtlib2_abstract_parser;
 
-void smtlib2_abstract_parser_init(smtlib2_abstract_parser *p,
-                                  smtlib2_context ctx);
-void smtlib2_abstract_parser_deinit(smtlib2_abstract_parser *p);
-void smtlib2_abstract_parser_parse(smtlib2_abstract_parser *p, FILE *src);
+#ifdef __cplusplus
+#define EXTERN_C       extern "C"
+#define EXTERN_C_BEGIN extern "C" {
+#define EXTERN_C_END   }
+#else
+#define EXTERN_C       /* Nothing */
+#define EXTERN_C_BEGIN /* Nothing */
+#define EXTERN_C_END   /* Nothing */
+#endif
+
+EXTERN_C_BEGIN
+	void smtlib2_abstract_parser_init(smtlib2_abstract_parser *p,
+			smtlib2_context ctx);
+	void smtlib2_abstract_parser_deinit(smtlib2_abstract_parser *p);
+	void smtlib2_abstract_parser_parse(smtlib2_abstract_parser *p, FILE *src);
+EXTERN_C_END
 
 #endif /* SMTLIB2ABSTRACTPARSER_H_INCLUDED */
